private/bufpkg/bufplugin/bufplugindocker/docker.go

```diff
@@ -0,0 +1,191 @@
+	if err != nil {
+		return nil, err
+	}
+
+	dockerContext, err := createDockerContext(dockerfile)
+	if err != nil {
+		return nil, err
+	}
+
+	eg, errGroupCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return buildkitSession.Run(errGroupCtx, func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
+			return d.cli.DialHijack(ctx, "/session", proto, meta)
+		})
+	})
+
+	buildID := stringid.GenerateRandomID()
+	imageName := pluginConfig.Name.IdentityString() + ":" + buildID
+	eg.Go(func() error {
+		defer buildkitSession.Close()
+
+		response, err := d.cli.ImageBuild(ctx, dockerContext, types.ImageBuildOptions{
+			Tags:       []string{imageName},
+			Dockerfile: "Dockerfile",
+			Platform:   "linux/amd64",
+			Labels: map[string]string{
+				"build.buf.plugins.config.remote": pluginConfig.Name.Remote(),
+				"build.buf.plugins.config.owner":  pluginConfig.Name.Owner(),
+				"build.buf.plugins.config.name":   pluginConfig.Name.Plugin(),
+			},
+			Version:   types.BuilderBuildKit, // DOCKER_BUILDKIT=1
+			SessionID: buildkitSession.ID(),
+		})
+		if err != nil {
+			return err
+		}
+		defer response.Body.Close()
+		scanner := bufio.NewScanner(response.Body)
+		for scanner.Scan() {
+			d.logger.Debug(scanner.Text())
+		}
+		if err := scanner.Err(); err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	imageInfo, _, err := d.cli.ImageInspectWithRaw(ctx, imageName)
+	if err != nil {
+		return nil, err
+	}
+	return &BuildResponse{
+		Image:  imageName,
+		Digest: imageInfo.ID,
+	}, nil
+}
+
+func (d *dockerAPIClient) Push(ctx context.Context, image string, auth RegistryAuth) (response *PushResponse, retErr error) {
+	registryAuth, err := auth.ToHeader()
+	if err != nil {
+		return nil, err
+	}
+	pushReader, err := d.cli.ImagePush(ctx, image, types.ImagePushOptions{
+		RegistryAuth: registryAuth,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, pushReader.Close())
+	}()
+	pushScanner := bufio.NewScanner(pushReader)
+	for pushScanner.Scan() {
+		d.logger.Debug(pushScanner.Text())
+	}
+	if err := pushScanner.Err(); err != nil {
+		return nil, err
+	}
+	return &PushResponse{}, nil
+}
+
+func (d *dockerAPIClient) Delete(ctx context.Context, image string) (*DeleteResponse, error) {
+	_, err := d.cli.ImageRemove(ctx, image, types.ImageRemoveOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return &DeleteResponse{}, nil
+}
+
+func (d *dockerAPIClient) Close() error {
+	return d.cli.Close()
+}
+
+type ClientOption func(client *dockerAPIClient)
+
+func NewClient(options ...ClientOption) (Client, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return nil, err
+	}
+	dockerClient := &dockerAPIClient{cli: cli}
+	for _, option := range options {
+		option(dockerClient)
+	}
+	if dockerClient.logger == nil {
+		dockerClient.logger = zap.L()
```

Comment:
If we require a `*zap.Logger` in the constructor so that we can always instrument it rather than using `zap.L`. This was mentioned in another comment above, but just mentioning it here for clarity.